pyadjoint/checkpointing.py

```diff
@@ -180,12 +185,19 @@ def _(self, cp_action, timestep):
                 for deps in self.tape.timesteps[timestep - 1].checkpointable_state:
                     self._global_deps.add(deps)
             else:
-                deps_to_clear = self._global_deps - self._global_deps.intersection(
-                    self.tape.timesteps[timestep - 1].checkpointable_state)
+                # Check if the block variables stored at `self._global_deps` are still
+                # dependencies in the previous timestep. If not, will remove them from the
+                # global dependencies.
+                deps_to_clear = self._global_deps.difference(self._global_deps.intersection(
+                    self.tape.timesteps[timestep - 1].checkpointable_state))
+
+                # Remove the block variables that are not global dependencies.
+                self._global_deps.difference_update(deps_to_clear)
+
+                # Clear the checkpoint once it is not a global dependency and should be stored
+                # only in the ``self.tape.timesteps`` checkpoints when needed.
```

Comment:
I'm afraid I don't quite understand what this means. Could you rephrase this?

Reply:
Better to understand this text: 
```
For no global dependencies, checkpoint storage occurs at a ``self.tape`` timestep only when required by an action from the schedule. Thus, we have to clear the checkpoint of block variables excluded from the ``self._global_deps``.

```

Reply:
Yeah I think that's good. Thanks.